wip: update gocode

anjmao opened the pull request:
This PR removes old gocode from nsf and adds gocode from mdempsky.  Any cache is not added yet.

keegancsmith commented:
thanks! Just ping when CI is happy and you want a review.

anjmao commented:
@keegancsmith 
* I fixed tests and added some todos for feature work since gocode doesn't support some of them like imports autocomplete, builtin global functions autocomplete...
* Some autocomplete tests are failing for go 1.8 since gocode dropped support for go 1.8 https://github.com/mdempsky/gocode/commit/4389b7c629e69da9a14ec72dc06f3f37fcea1e00 Should we consider dropping go 1.8 support in langserver (make sense to me)?

keegancsmith commented:
Yup dropping support of go1.8 sounds good to me. If I am not mistaken most tooling aims to support the last 2 minor releases. Do you mind changing the travis config to be `"1.10.x"` and `"1.11.x"` in this PR?

keegancsmith commented:
Finally took a look at the PR. You are removing functionality that was added in that is relied on, like `SetBuildContext` and `InitDaemon`. You should just be able to keep `export.go`, I believe that was added to export the functionality we hook into.

I also see there are lots of tests you have now commented out. Anyidea why they are breaking? Could it be due to the change in `SetBuildContext`, or is it just due to how much this fork of gocode has diverged from upstream?

anjmao commented on langserver/completion.go:
```diff
@@ -46,11 +47,20 @@ func (h *LangHandler) handleTextDocumentCompletion(ctx context.Context, conn jso
+		Data:     contents,
+		Cursor:   offset,
+		Source:   !h.config.UseBinaryPkgCache,
+		Context:  gbimporter.PackContext(h.BuildContext(ctx)),
```
@keegancsmith Thanks for review. 

As I'm not using gocode daemon (not starting any net listener) I'm setting build context here. In current master https://github.com/sourcegraph/go-langserver/blob/master/langserver/internal/gocode/export.go#L17 is used only in unit tests. 
There are failing integration tests since this gocode fork returns different suggestions for some cases.

keegancsmith replied:
Is it possible to update those tests instead of just commenting them out?

Ok great about setting the buildcontext where you set it, that actually makes a lot of sense.

keegancsmith commented:
If you could update the failing tests that would be great, then I'll merge in.

anjmao commented:
Sure I will luck at these failing tests, but some tests are failing since gocode doesn't support imports autocomplete, I thought it is better if I add this feature on separate PR or do you want me to add it in this PR?

keegancsmith commented:
Another PR works for me.

anjmao commented on langserver/completion.go:
```diff
@@ -46,11 +47,21 @@ func (h *LangHandler) handleTextDocumentCompletion(ctx context.Context, conn jso
+		Filename: filename,
+		Data:     contents,
+		Cursor:   offset,
+		Builtin:  true,
```
@keegancsmith I was missing this flag, after adding `Builtin: true` I was able to fix failing tests except whos with package autocomplete. There is one bug in gocode , if I place dot inside package path I get all builtins. In real world this is not a big deal and this can also be fixed with package autocomplete.

![image](https://user-images.githubusercontent.com/3028012/45281426-6da4f000-b4e0-11e8-8585-275c30c2313f.png)

anjmao commented:
@keegancsmith Can you have a look now, I added Builtin: true flag and updated some tests which was not working before.

anjmao commented on langserver/langserver_test.go:
```diff
@@ -1428,13 +1426,15 @@ func lspTests(t testing.TB, ctx context.Context, h *LangHandler, c *jsonrpc2.Con
+
+		h.config.UseBinaryPkgCache = false
+
+		// TODO(anjmao): autocomplete tests fails if binary cache is used
```
As gocode is using binary cache for completion if I set this to UseBinaryPkgCache = true a lot of  completion tests are failing since autocomplete returns empty result because test case binary is not built yet at that moment. Is there any way I can pre build each test case code during mountFS?

lloiser replied:
The tests already install the packages:

https://github.com/sourcegraph/go-langserver/blob/68907e07ef2b8479a592485f63a225c36e23ed67/langserver/langserver_test.go#L1401-L1409

Maybe there is another problem that I cannot see at the moment...

slimsag replied:
If I understand correctly, this is only failing because `$GOPATH/pkg` would be used when `UseBinaryPkgCache = true`, and in the case of the tests here we are not actually building the code with `go install` because the code is not on disk. i.e. this wouldn't actually fail in practice / in real use when `UseBinaryPkgCache = true`, yeah?

If so, I think it is fine to not address this for now.

slimsag replied:
Yeah, I am wrong here (I submitted my reply before I saw @lloiser's reply). We are copying the test sources into a temporary GOPATH and `go install`'ing them, so it should be working

anjmao replied:
@lloiser Thanks, yes it's actually building code and pkg binary cache files are created during tests, but for some reasons imported packages scope is empty. 

In gocode package imports prints empty scope if binary cache is used, but it works perfectly fine in vscode when using binary cache.
```
pkg.Imports()[0].Scope()
```
I'm currently investigating `go/importer` used in gocode.

anjmao commented on .travis.yml:
```diff
@@ -1,7 +1,6 @@
 go:
-  - 1.8.x
   - 1.9.x
   - 1.10.x
```
Yep, good point. After adding 1.11 I needed to adjust one test since @ is not valid import path anymore.

anjmao commented:
@lloiser @keegancsmith I finally found why completion tests was not working with BinaryCache.

1. Since I'm on macOS my GOOS is darwin, tests was using hardcoded linux. I changed tests to use runtime.GOOS instead.
2. Gocode has support for gb pased projects and there was a bug with importPaths, which I fixed in separate commit.

I never used https://getgb.io/, but I'm not sure if we should support that. I would rather drop gb support since custom GOPATH is a easy solution to someone who want to have project based workflow.
